File: packages/rag/prompting.py
from __future__ import annotations

from packages.rag.citations import format_citations
from packages.rag.retrieval import RetrievedChunk


def build_grounded_prompt(question: str, chunks: list[RetrievedChunk]) -> str:
    if not chunks:
        return question

    context_lines: list[str] = []
    citations = format_citations(chunks)
    for chunk, citation in zip(chunks, citations):
        context_lines.append(f"{citation} {chunk.content}")

    context = "\n".join(context_lines)
    return (
        "Answer the question using only the provided context. "
        "Include citations using the provided source tags.\n\n"
        f"Context:\n{context}\n\nQuestion:\n{question}"
    )

File: internal/rag/prompting.go
package rag

import "strings"

func BuildGroundedPrompt(question string, chunks []RetrievedChunk) string {
	if len(chunks) == 0 {
		return question
	}

	contextLines := make([]string, 0, len(chunks))
	citations := FormatCitations(chunks)
	for i, chunk := range chunks {
		contextLines = append(contextLines, citations[i]+" "+chunk.Content)
	}

	context := strings.Join(contextLines, "\n")
	return "Answer the question using only the provided context. " +
		"Include citations using the provided source tags.\n\n" +
		"Context:\n" + context + "\n\nQuestion:\n" + question
}
